code/code/sys/process.cc

```diff
@@ -195,22 +192,23 @@ void TScheduler::runChar(int pulseNum)
           tmp_ch->desc && tmp_ch->desc->character != tmp_ch ? "bad char ptr in desc" :
           "bug"
           ));
-      tmp_ch = temp;
-      temp = temp->next;
+      tmp_ch = tmp_ch->next;
+      continue;
     }
 
     if(!count--)
       break;
 
     if (tmp_ch->getPosition() == POSITION_DEAD) {
-      assert(tmp_ch); // I suppose this guy exists if we have already dereferenced it
       // even if shortDescr is NULL.
       vlogf(LOG_BUG, format("Error: dead creature (%s at %d) in character_list, removing.") % 
           tmp_ch->getName() % tmp_ch->in_room);
-      delete tmp_ch;
-      tmp_ch = temp;
+      TBeing* dead = tmp_ch;
+      tmp_ch = tmp_ch->next;
```

Comment:
So we move the tmp_ch pointer to next but don't we need to move the previous items next pointer to remove it from the list? It's been a long time since I thought about linked lists maybe there's some magic here that I don't understand.

Reply:
<img width="559" alt="Screen Shot 2021-02-25 at 9 41 02 PM" src="https://user-images.githubusercontent.com/1966964/109260023-3c086800-77b2-11eb-84a0-66c17861a883.png">
I found a pic online in case my last comment didn't make sense.

Reply:
In the old code, there was a pointer to current (`tmp_ch`) and a backup pointer to the next one, in case we need to delete the current one. That was the bug - in case the current proc deletes the next mob in the list, the next pointer would point to a deleted mob.

In the new code, we just have one pointer to current, `tmp_ch`, and deal with deletions as they come (advance through the list, stop processing this one).

Reply:
But don't we need to maintain the linked list by updating the "next" pointer of the previous item in the list?

Reply:
As weird as it sounds - no, as the destructor of TBeing does that.

Reply:
Ahh ok. That makes sense. So it should be safe unless we point to next *before* we delete next as was the case previously.